streaming/base/dataset/streaming.py

```diff
@@ -0,0 +1,639 @@
+        raw_filename = os.path.join(self.local, self.split, raw_info.basename)
+        if os.path.isfile(raw_filename):
+            return
+
+        # Is compression used?
+        if zip_info:
+            # Download the compressed form if missing.
+            zip_filename = os.path.join(self.local, self.split, zip_info.basename)
+            if not os.path.isfile(zip_filename):
+                self._download_file(zip_info.basename)
+
+            # Validate and decompress.
+            self._decompress_shard_part(zip_info, zip_filename, raw_filename, compression)
+        else:
+            # Download the raw form.
+            self._download_file(raw_info.basename)
+
+            # Validate if requested.
+            if self.validate_hash:
+                data = open(raw_filename, 'rb').read()
+                assert get_hash(self.validate_hash, data) == raw_info.hashes[self.validate_hash]
+
+    def _download_shard(self, shard_id: int) -> None:
+        """Download the given shard.
+
+        Args:
+            shard_id (int): Shard ID.
+        """
+        reader = self.shards[shard_id]
+        for raw_info, zip_info in reader.file_pairs:
+            self._download_shard_part(shard_id, raw_info, zip_info, reader.compression)
+
+    def _download_or_await_shard(self, lock: FileLock, shard_states: NDArray[np.uint8],
+                                 shard_id: int) -> None:
+        """Either download the given shard or wait on its download.
+
+        Args:
+            lock (FileLock): The lock protecting ``shard_states``.
+            shard_states (NDArray[np.uint8]): The download status of each shard, as an array in
+                shared memory.
+            shard_id (int): Shard ID.
+        """
+        # First, the fast path: check the shared memory shard state without taking the lock. The
+        # shard states only ever go up, so if we're at the downloaded state, it's downloaded.
+        state = shard_states[shard_id]
+        if state == _ShardState.DOWNLOADED:
+            return
+
+        # Shard is not necessarily downloaded, so check and update state with the lock.
+        lock.acquire()
+        state = shard_states[shard_id]
+        if state == _ShardState.UNKNOWN:
+            shard_states[shard_id] = _ShardState.DOWNLOADING
+            lock.release()
+            self._download_shard(shard_id)
+            with lock:
+                shard_states[shard_id] = _ShardState.DOWNLOADED
+        elif state == _ShardState.DOWNLOADING:
+            lock.release()
+            while shard_states[shard_id] != _ShardState.DOWNLOADED:
+                sleep(0.07)
+        elif state == _ShardState.DOWNLOADED:
+            lock.release()
+        else:
+            raise RuntimeError('Unknown shard state')
+
+    def _download_thread(self, epoch: int, sample_ids: NDArray[np.int64]) -> None:
+        """Download the relevant shards in the background while we are being iterated.
```

Comment:
Wondering how many threads will there be running at once? For example, let say, the model training is happening on 1 GPU with `num_workers` as `8` in `torch.utils.data.DataLoader`. Does each worker process has 1 thread running to download the shards? or 1 node has only 1 thread running to download the shards?

Can you please add the comment saying how many thread would be running at a same time in avg.case?

Reply:
there's one download thread per worker, which is invoked at the start of the epoch (because even if we are persistent, the shards to download this worker could conceivably change for some reason), and exits as soon as it detects a new epoch has started (because only one epoch can be valid at a time)

Reply:
we use a thread here because the main thread is busy iterating which sample to yield, while the thread updates which sample to download

in a perfect world, it might be faster to have it skip samples that some other thread is already downloading, so you'd track three indices: sample to yield, last sample that is downloaded, and sample that we are downloading

Reply:
we would tune this to optimize the number of concurrent shard downloads to maximize throughput, which is backend-dependent, say local filesystem is probably approaching 1

Reply:
Updated docstring?